### PR TITLE
add stripPrefix and Pattern for Data.List

### PR DESCRIPTION
**Pull request by safareli**
https://github.com/purescript-contrib/purescript-parsing/pull/62#discussion_r119991248

### CODE REVIEW

**Comment by safareli** [comment]
if you know a better place in the file for stripPrefix and Pattern let me know

File: src/Data/List.purs [hunk 1]
```diff
@@ -476,6 +479,32 @@ sortBy cmp = mergeAll <<< sequences
 -- Sublists --------------------------------------------------------------------
 --------------------------------------------------------------------------------
 
+-- | A newtype used in cases where there is a list to be matched.
+newtype Pattern a = Pattern (List a)
+
+derive instance eqPattern :: (Eq a) => Eq (Pattern a)
+derive instance ordPattern :: (Ord a) => Ord (Pattern a)
+derive instance newtypePattern :: Newtype (Pattern a) _
```
Comment (safareli) [inline src/Data/List.purs]: not sure if this instance is correct

Comment (garyb) [inline src/Data/List.purs]: Yep, that's how deriving for `Newtype` works.

File: src/Data/List.purs [hunk 2]
```diff
@@ -476,6 +479,32 @@ sortBy cmp = mergeAll <<< sequences
...
+-- | * `stripPrefix (Pattern (1:Nil)) (1:2:Nil) == Just (2:Nil)`
+-- | * `stripPrefix (Pattern Nil) (1:Nil) == Just (1:Nil)`
+-- | * `stripPrefix (Pattern (2:Nil)) (1:Nil) == Nothing`
+-- |
+-- | Running time: `O(n)` where `n` is the number of elements to strip.
+stripPrefix :: forall a. Eq a => Pattern a -> List a -> Maybe (List a)
+stripPrefix (Pattern p') s = tailRecM2 go p' s
+  where
+  go prefix input = case prefix, input of
+    (Cons p ps), (Cons i is) | p == i -> Just $ Loop ({ a: ps, b: is })
```
Comment (paf31) [inline src/Data/List.purs]: The parens are redundant in the patterns and in the value here.

File: src/Data/List.purs [hunk 3]
```diff
@@ -476,6 +479,32 @@ sortBy cmp = mergeAll <<< sequences
 -- Sublists --------------------------------------------------------------------
 --------------------------------------------------------------------------------
 
+-- | A newtype used in cases where there is a list to be matched.
+newtype Pattern a = Pattern (List a)
+
+derive instance eqPattern :: (Eq a) => Eq (Pattern a)
```
Comment (paf31) [inline src/Data/List.purs]: Redundant parens.

**Comment by paf31** [comment]
👍 Thanks!

**Comment by paf31** [comment]
@garyb Do you want to use the `Pattern` newtype here too?

This looks good to me.

**Comment by garyb** [comment]
The name `Pattern` is maybe a little strange in the context of lists, but I do think it's better to newtype with something so that the argument order is unambiguous. I can't think of a better name either, so 👍 I guess!

**Comment by paf31** [comment]
`Prefix`?

**Comment by garyb** [comment]
That works for sure, but if there's a chance any of the other string-ish functions will be replicated here then `Pattern` is more widely applicable. Even if just `stripSuffix` was added. :wink:

**Comment by paf31** [comment]
True ... ok, merging.

**Pull request merged**